text/0004-type-deduction.md

```diff
@@ -0,0 +1,214 @@
+
+## As part of a binary expression
+A distinction has to be made by the operator in use (and depending on the operator additionaly the side on which the variable appears).
+
+### Operator && || (Logical Operators)
+```shell Example
+ls | where $it.name == LICENSE || $catch_all
+$catch_all -> SyntaxShape::Boolean
+```
+The variable can be of any type that is automaticaly decayable to a boolean value.
+
+### Operator In NotIn
+#### Variable on right side
+```shell Example
+ls | where name in $values
+$values -> SyntaxShape::Table
+```
+### Operator In NotIn
+#### Variable on left side
+```shell Example
+ls | where $value in [...]
+$values -> All SyntaxShapes present in the Table ([...])
+```
+
+### Operator Plus Minus
+```shell Example 1
+ls | where size < 1kb + $offset
+$offset -> SyntaxShape::Unit
+```
+```shell Example 2
+ls | where 1.5 < 1 + $var
+$var -> SyntaxShape::Number, SyntaxShape::Int
+```
+The shape of the variable is Unit if other side of the binary expression is Unit, Number or Int otherwise.
+
+### Operator Multiply Divide
+```shell Example
+ls | where size < 1.5 * $size
+$size -> SyntaxShape::Unit
+```
+In general: the variable can be one of Int, Number, Unit.
+The variable can't be Unit if the result expression has a undefined unit type (e.G. Unit * Unit or Int / Unit gives an undefined unit type for every unit nu currently implements).
+The variable must be of Unit type, if the result type of the binary expression is used as a Unit type and the not variable side of the binary is a Number or Unit (see example above).
+
+### Operator Contains NotContains
+The variable must be of type String.
+
+
+## Correct Inference for dependencies
+Extra care has to be taken when deducing in binary expressions, with an Operator (like Plus, Minus, Multiply, Divide, In (var on lhs), NotIn (var on lhs)) where the variable type depends on the type of the other side. There are 2 special cases that have to be handled.
+
+### Special case 1: Binary with variables on both sides
+```shell Example
+ls | where size < $a * $b; kill $a
+$a -> SyntaxShape::Int
+$b -> SyntaxShape::Unit
+```
+
+At the time of traversing the AST, any deduction for any variable may not be present, and if present may not be complete. Therefore constellations in which the deduction of one variable depends on the deduction of another one (we will call these constellations 'dependencies'), have to be postponed. 
+
+### Special case 2: Binary which depends on the result type of another expression
+```shell Example
+ls | where size < $a * ($b * $c); kill $b $c
+$a -> SyntaxShape::Unit
+$b -> SyntaxShape::Int
+$c -> SyntaxShape::Int
+```
+Same as above. But this time $a also depends on the result type of ($b * $c). This dependency has to be postponed, as $b nor $c can be deduced at the point in time of traversing the AST.
```

Comment:
This essentially feels like `Special case 1`. Basically, you have `a <op> b`, `a` and `b` both being a math expression (could be a number, could be another binary operation) and you can't deduce the type for both `a` and `b`.

I'm wondering if we could combine the two, to simplify the RFC?